mantidproject/mantid
# Elemental Analysis Plotting Wrapper/Interface

**Description of work.**
Adds a plotting window (interface) that:
- allows 'dynamic' adding and automatic resizing of 1-4 plots
- has a `ComboBox` for plot selection (when changing plot limits)
- has x and y limit changing
- exposes methods to:
    - add vertical lines
    - add horizontal lines
    - plot workspaces (with/without errors)
    - add/remove subplots
    - ~add moveable lines~ (future PR)
- is unit tested (AxisChanger)
- ~is unit tested (Plotting)~ (#23325)

*The Periodic Table functionality has not changed*

 ---
**Notes**
When `Errors` is checked, all the subplots must be cleared then redrawn as `matplotlib` exposes no way to remove a plot from a subplot. If a title/lines were added, then these should be redrawn after `clear()` is called. For this, the additions must be 'saved' to be replayed when the subplots are being redrawn. An additional method is exposed to 'save' arbitrary methods.

The `Add`/`Remove` buttons on the Elemental Analysis window are temporary:
- `Add` will add a plot to the plotting window
- `Remove` will remove a random plot from the plotting window

Once closed, the plotting window can only be reopened by opening a new Elemental Analysis window: this is by design and will change in a future PR

---

**Report to:** Aidy

---
**To test:**
- `Interfaces->Muon->Elemental Analysis`
- All plots should consist of a sample workspace, the result of an addition of two sample workspaces and have a horizontal line at `y = 0.06`, a vertical line at `x = 10100` and a title.
- When plots are added/deleted, the `ComboBox` will update to contain all plots.
- When `Add` is clicked, the limits will stay the same, as the selected plot is the first one in the list.
- All plot elements (title, lines, workspaces, limits) should remain on add/del (unless the plot is deleted)
- Toggling `Errors` will update all plots with error bars, new limits (if necessary), with the plot elements remaining on the plot (this may take a couple of seconds - see *Notes*)
- Typing in new limits (pressing `Return` to update) will update the specified plot with the new limits (the graph can be flipped in the x/y planes if necessary)
- A new plot should automatically have its current limits entered in the axis changer boxes
- An `IndexError` will be thrown if you try and delete a plot where none exist (this is the example delete function in `Elemental Analysis`, else a `KeyError` would be thrown (intentional))
- A `KeyError` will be thrown if a user tries to add a fifth plot

---
Closes #23260

*This does not require release notes* because **they were added in a previous PR**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
